### PR TITLE
Added content and functionality for Allowable Losses 

### PR DESCRIPTION
**Pull request by macgrewal**

### CODE REVIEW

File: app/common/KeystoreKeys.scala [hunk 1]
```diff
@@ -50,6 +50,7 @@ trait KeystoreKeys {
     val acquisitionValue = "res:acquisitionValue"
     val acquisitionCosts = "res:acquisitionCosts"
     val improvements = "res:improvements"
+    val allowableLosses = "res:improvements"
```
Comment (DavidWoodward37) [inline app/common/KeystoreKeys.scala]: Should this key not be res:allowableLosses?

Comment (macgrewal) [inline app/common/KeystoreKeys.scala]: Yes

File: app/controllers/resident/DeductionsController.scala [hunk 2]
```diff
@@ -124,8 +122,26 @@ trait DeductionsController extends FeatureLock {
...
-  val allowableLosses = Action.async { implicit request =>
-    Future.successful(Ok(views.allowableLosses()))
+  val allowableLosses = FeatureLockForRTT.async { implicit request =>
+    calcConnector.fetchAndGetFormData[AllowableLossesModel](KeystoreKeys.ResidentKeys.allowableLosses).map {
+      case Some(data) => Ok(views.allowableLosses(allowableLossesForm.fill(data)))
+      case None => Ok(views.allowableLosses(allowableLossesForm))
+    }
+  }
+
+  val submitAllowableLosses = Action.async { implicit request =>
```
Comment (DavidWoodward37) [inline app/controllers/resident/DeductionsController.scala]: Should this not be the FearureLockForRTT rather than the Action?

Comment (macgrewal) [inline app/controllers/resident/DeductionsController.scala]: Yes

File: test/controllers/resident/DeductionsControllerTests/AllowableLossesActionSpec.scala [hunk 3]
```diff
@@ -16,29 +16,125 @@
...
+    "request has a valid session and no keystore value" should {
+
+      lazy val target = setupTarget(None)
+      lazy val result = target.allowableLosses(fakeRequestWithSession)
+
+      "return a status of 200" in {
+        status(result) shouldBe 200
+      }
+
+      s"return some html with " in {
```
Comment (DavidWoodward37) [inline test/controllers/resident/DeductionsControllerTests/AllowableLossesActionSpec.scala]: This is being really really really picky but I don't think you need that -> s" at the start of the string as you're not using any variables.

File: test/controllers/resident/DeductionsControllerTests/AllowableLossesActionSpec.scala [hunk 4]
```diff
@@ -16,29 +16,125 @@
...
+      lazy val result = target.allowableLosses(fakeRequestWithSession)
+
+      "return a status of 200" in {
+        status(result) shouldBe 200
+      }
+
+      s"return some html with title of ${messages.title}" in {
+        contentType(result) shouldBe Some("text/html")
+        Jsoup.parse(bodyOf(result)).title shouldEqual messages.title
+      }
```
Comment (DavidWoodward37) [inline test/controllers/resident/DeductionsControllerTests/AllowableLossesActionSpec.scala]: Should you split this into the two tests as in the test above?

File: test/controllers/resident/DeductionsControllerTests/AllowableLossesActionSpec.scala [hunk 5]
```diff
@@ -16,29 +16,125 @@
...
+
+    "a valid form 'Yes' is submitted" should {
+      lazy val request = fakeRequestToPOSTWithSession(("isClaiming", "Yes"))
+      lazy val result = DeductionsController.submitAllowableLosses(request)
+
+      "return a 303" in {
+        status(result) shouldBe 303
+      }
+
+      "redirect to the reliefs entry page" in {
```
Comment (DavidWoodward37) [inline test/controllers/resident/DeductionsControllerTests/AllowableLossesActionSpec.scala]: "redirect to the allowable losses entry page" ?

File: test/controllers/resident/DeductionsControllerTests/AllowableLossesActionSpec.scala [hunk 6]
```diff
@@ -16,29 +16,125 @@
...
+
+    "a valid form 'No' is submitted" should {
+      lazy val request = fakeRequestToPOSTWithSession(("isClaiming", "No"))
+      lazy val result = DeductionsController.submitAllowableLosses(request)
+
+      "return a 303" in {
+        status(result) shouldBe 303
+      }
+
+      "redirect to the other properties page" in {
```
Comment (DavidWoodward37) [inline test/controllers/resident/DeductionsControllerTests/AllowableLossesActionSpec.scala]: "redirect to the losses-brought-forward page"?

File: test/controllers/resident/DeductionsControllerTests/AllowableLossesActionSpec.scala [hunk 7]
```diff
@@ -16,29 +16,125 @@
...
+      lazy val result = DeductionsController.submitAllowableLosses(request)
+      lazy val doc = Jsoup.parse(bodyOf(result))
+
+      "return a 400" in {
+        status(result) shouldBe 400
+      }
 
-    s"return a title of ${messages.title}" in {
-      Jsoup.parse(bodyOf(result)).title shouldEqual messages.title
+      "render the reliefs page" in {
```
Comment (DavidWoodward37) [inline test/controllers/resident/DeductionsControllerTests/AllowableLossesActionSpec.scala]: "render the allowable losses page"?

File: test/forms/resident/AllowableLossesFormSpec.scala [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    "supplied with invalid data" should {
+      lazy val form = allowableLossesForm.bind(Map(("isClaiming", "a")))
+
+      "return a form with errors" in {
+        form.hasErrors shouldBe true
+      }
+
+      s"return an error with message ${commonMessages.undefinedMessage}" in {
+        form.error("isClaiming").get.message shouldBe commonMessages.undefinedMessage
+      }
```
Comment (DavidWoodward37) [inline test/forms/resident/AllowableLossesFormSpec.scala]: Do you also want to include a check for the length of the error array? Just to make sure only one has been caught?

**Pull request merged**